compiler/rustc_middle/src/ty/mod.rs

```diff
@@ -1609,44 +1609,66 @@ pub type PlaceholderType = Placeholder<BoundVar>;
 pub type PlaceholderConst = Placeholder<BoundVar>;
 
 /// A `DefId` which is potentially bundled with its corresponding generic parameter
-/// in case `did` is a const argument.
-///
-/// This is used to prevent cycle errors during typeck
-/// as `type_of(const_arg)` depends on `typeck(owning_body)`
-/// which once again requires the type of its generic arguments.
-///
-/// Luckily we only need to deal with const arguments once we
-/// know their corresponding parameters. We (ab)use this by
-/// calling `type_of(param_did)` for these arguments.
+/// in case `did` is a const argument. Used to prevent cycle errors during `typeck`.
 ///
 /// ```rust
 /// #![feature(const_generics)]
 ///
 /// struct A;
 /// impl A {
-///     fn foo<const N: usize>(&self) -> usize { N }
+///     fn foo<const N: usize>(&self) -> [u8; N] { [0; N] }
+///     //           ^ const parameter
 /// }
 /// struct B;
 /// impl B {
-///     fn foo<const N: u8>(&self) -> usize { 42 }
+///     fn foo<const M: u8>(&self) -> usize { 42 }
+///     //           ^ const parameter
 /// }
 ///
 /// fn main() {
 ///     let a = A;
-///     a.foo::<7>();
+///     let _b = a.foo::<7>();
+///     //               ^ const argument
 /// }
 /// ```
+///
+/// Let's look at the call `a.foo::<7>()` here. We do not know which
+/// `foo` is used until we know the type of `a`.
+///
+/// We only know the type of `a` once we are inside of `typeck(main)`.
+/// At that point we also want to know the type of `_b` which depends on
+/// the value of the const argument.
```

Comment:
Should "value" be "type"?

Reply:
Oh, is that the entire point? Naively it might seem like we need the value, but really we only need its type for now, and to determine the type we only need to figure out the const parameter it instantiates?

Reply:
I still don't understand why the *type* of `_b` depends on the *value* of the const argument. In the example, it does not. The type of `_b` is determined once the right instance of `foo` is picked.

Or is the point that the type could be `[u8; N]` where the `N` is the const argument? I figured we could just reference the unevaluated argument without running any query, but maybe that does not work.